gnomad_constraint/pipeline/constraint_pipeline.py

```diff
@@ -0,0 +1,158 @@
+"""
+This script builds a constraint pipeline that calculates constraint metrics.
+
+The constraint metrics pipeline will compute for LoF variants, missense variants, and synonymous variants:
+    - The number of observed variants
+    - The number of expected variants
+    - The observed: expected ratio
+    - The confidence interval around the observed: expected ratio
+    - pLI score (Probability of loss-of-function intolerance; probability that transcript falls into distribution of haploinsufficient genes)
+    - pNull (Probability that transcript falls into distribution of unconstrained genes)
+    - pRec (Probability that transcript falls into distribution of recessive genes)
+    - z-scores (Measure how constrained or intolerant a gene or transcript is to missense variants and synonymous variants)
+
+The constraint pipeline consists of the following parts:
+    - preprocess data
+    - create training set
+    - build models
+    - apply models
+    - calculate metrics
+"""
+
+import argparse
+import logging
+
+import hail as hl
+from gnomad.utils.filtering import filter_x_nonpar, filter_y_nonpar
+from gnomad.utils.reference_genome import get_reference_genome
+
+from gnomad_constraint.resources.resource_utils import (
+    CURRENT_VERSION,
+    DATA_TYPES,
+    VERSIONS,
+    annotated_context_ht,
+    get_logging_path,
+    get_preprocessed_ht,
+    get_sites_resource,
+)
+from gnomad_constraint.utils.constraint import (
+    add_vep_context_annotations,
+    prepare_ht_for_constraint_calculations,
+)
+
+logging.basicConfig(
+    format="%(asctime)s (%(name)s %(lineno)s): %(message)s",
+    datefmt="%m/%d/%Y %I:%M:%S %p",
+)
+logger = logging.getLogger("constraint_pipeline")
+logger.setLevel(logging.INFO)
+
+
+def main(args):
+    """Execute the constraint pipeline."""
+    hl.init(log="/constraint_pipeline.log")
+
+    test = args.test
+    overwrite = args.overwrite
+    # TODO: gnomAD v4 is still in production, for now this will only use 2.1.1.
+    version = args.version
+    if version not in VERSIONS:
+        version = CURRENT_VERSION
+        logger.warning(
+            "The requested version of resource Tables are not exist, will use gnomAD"
+            " v2.1.1 as default."
+        )
+
```

Comment:
Just a thought, maybe add this here, and then use in the way I suggest below and in the training dataset PR.
```suggestion
    preprocess_resources = {}
    training_resources = {}
    for region in GENOMIC_REGIONS:
        for data_type in DATA_TYPES:
            if (region == "autosome_par") | (data_type != "genomes"):
                preprocess_resources[(region, data_type)] = get_preprocessed_ht(
                    data_type, version, region, test
                )
        training_resources[region] = get_training_dataset(version, region, test)
```